### PR TITLE
Pre-Compile Templates for Better Performance

### PR DESCRIPTION
**Pull request by kaendfinger**
Pre-Compiling Templates gives a decent speed boost.

### CODE REVIEW

**Comment by kaendfinger** [comment]
@keertip

File: lib/src/html_generator.dart [hunk 1]
```diff
@@ -30,21 +30,21 @@ class HtmlGenerator extends Generator {
   final String generatedOn;
 
   // TODO: fix this to be a single map, just like _partialTemplates
-  static final String indexTemplate = _loadTemplate('templates/index.html');
-  static final String libraryTemplate =
+  static final Function indexTemplate = _loadTemplate('templates/index.html');
```
Comment (sethladd) [inline lib/src/html_generator.dart]: Can we use a typedef here? Function is so... non-descriptive :)

File: lib/src/html_generator.dart [hunk 2]
```diff
@@ -204,21 +204,18 @@ class HtmlGenerator extends Generator {
     return f;
   }
 
-  void _writeFile(String filename, String template, Map data) {
+  void _writeFile(String filename, template, Map data) {
```
Comment (sethladd) [inline lib/src/html_generator.dart]: Use type annotation for template

Comment (kaendfinger) [inline lib/src/html_generator.dart]: Unable to. We pass both strings and functions to it.

**Comment by kaendfinger** [comment]
@sethladd this saves about 4 seconds when generating the SDK docs on my MacBook Pro.

**Comment by kaendfinger** [comment]
@sethladd ping

**Comment by sethladd** [comment]
Thanks for working on the performance fix. I'm very uneasy about a function that returns dynamic. Can you do a deeper refactoring so that each function returns exactly one type ?

**Comment by kaendfinger** [comment]
@sethladd Sure

**Comment by kaendfinger** [comment]
I'm going to actually redo this PR so you don't have to fix the conflicts.

**Pull request closed without merging**